src/main/java/com/code_intelligence/jazzer/mutation/mutator/lang/FloatingPointMutatorFactory.java

```diff
@@ -40,31 +52,203 @@ public Optional<SerializingMutator<?>> tryCreate(AnnotatedType type, MutatorFact
+            break;
+          default: // random in range cannot exceed the given bounds (and cannot be NaN)
+            return (prng.closedRange(minValue, maxValue));
+        }
+      }
+      result = forceInRange(result, minValue, maxValue, allowNaN);
+      // If value repeats, or is a not allowed NaN, mutate again
+      if (result == value || (Float.isNaN(result) && Float.isNaN(value))
+          || (Float.isNaN(result) && !allowNaN)) {
+        return mutate(value, prng);
+      }
+      return result;
+    }
+
+    static float forceInRange(float value, float minValue, float maxValue, boolean allowNaN) {
+      if ((value >= minValue && value <= maxValue) || (Float.isNaN(value) && allowNaN)) {
+        return value;
+      }
+
+      // Clamp infinite values
+      if (value == Float.POSITIVE_INFINITY)
+        return maxValue;
+      if (value == Float.NEGATIVE_INFINITY)
+        return minValue;
+
+      // If NaN was allowed, it was handled above.
+      // Here we replace NaN by the middle of the clamped finite range.
+      if (Float.isNaN(value)) {
+        // maxValue or minValue may be infinite, so we need to clamp them.
+        return minValue
+            + (Math.min(Float.MAX_VALUE, maxValue) * 0.5f
+                - Math.max(Float.MIN_VALUE, minValue) * 0.5f);
+      }
+
+      // By now all infinity / NaN cases should have been handled
+      // the results from here on out should be finite.
+      float finiteMax = Math.min(Float.MAX_VALUE, maxValue);
+      float finiteMin = Math.max(-Float.MAX_VALUE, minValue);
+      float range = finiteMax - finiteMin;
+
+      if (range == 0f)
+        return finiteMin;
+
+      float diff = value - finiteMin;
+
+      if (Float.isFinite(diff) && Float.isFinite(range)) {
+        return finiteMin + Math.abs(diff % range);
+      }
+
+      // diff, range, or both are infinite: divide both by 2, reduce, and multiply by 2.
+      float halfDiff = value * 0.5f - finiteMin * 0.5f;
+      return finiteMin + (halfDiff % (finiteMax * 0.5f - finiteMin * 0.5f)) * 2.0f;
+    }
+
+    public float mutateWithMathematicalFn(float value, PseudoRandom prng) {
+      double result = prng.pickIn(mathFunctions).apply((double) value);
+      return (float) result;
+    }
+
+    private float mutateWithBitFlip(float value, PseudoRandom prng) {
+      int bits = Float.floatToRawIntBits(value);
+      int bitToFlip = prng.closedRange(0, 31);
+      bits ^= 1L << bitToFlip;
+      return Float.intBitsToFloat(bits);
+    }
+
+    private float mutateExponent(float value, PseudoRandom prng) {
+      int bits = Float.floatToRawIntBits(value);
+      int exponent = (bits >> 23) & 0xff + prng.closedRange(0, EXPONENT_RANDOM_WALK_RANGE);
+      bits = (bits & ~(0xff << 23)) | ((exponent % 0xff) << 23);
+      return Float.intBitsToFloat(bits);
+    }
+
+    private float mutateMantissa(float value, PseudoRandom prng) {
+      int bits = Float.floatToRawIntBits(value);
+      int mantissa = bits & 0x7fffff;
```

Comment:
Let's pull out the numeric value into a `MANTISSA_MASK` constant.

Reply:
Good idea, I updated the DoubleMutator as well.